Lacrymology/new caspa integration

Lacrymology opened the pull request:
integrate the changes in playlists / clips models from the caspa rewrite

Lacrymology committed:
pieces now come from a different mongo collection

Signed-off-by: Tomas Neme <lacrymology@gmail.com>

Lacrymology committed:
fps now is in piece.video, not directly under piece

Signed-off-by: Tomas Neme <lacrymology@gmail.com>

Lacrymology committed:
don't use Date

Signed-off-by: Tomas Neme <lacrymology@gmail.com>

Lacrymology committed:
use $ instead of - for playlist/clip id separators in filename

Signed-off-by: Tomas Neme <lacrymology@gmail.com>

Lacrymology commented on utils.js:
```diff
@@ -2,11 +2,11 @@ var moment = require('moment');
 exports = module.exports = {
     getXmlFileNameFromClip: function(clip) {
-        return clip.playlist_id + "-" + clip.id + ".xml";
+        return clip.playlist_id + "$" + clip.id + ".xml";
```
this is horrible, but it's legacy-horrible, and I don't know how to avoid it

jmrunge commented:
Travis:

``` javascript
1 of 193 tests failed:

  1) PlaylistMongoDriver #subscriptions should respond to update messages:
     TypeError: Cannot read property 'pieces' of null
      at /home/travis/build/inaes-tic/mbc-mosto/drivers/playlists/mongo-driver.js:6:8093
      at /home/travis/build/inaes-tic/mbc-mosto/node_modules/mbc-common/node_modules/mongoskin/node_modules/mongodb/lib/mongodb/collection.js:957:5
      at /home/travis/build/inaes-tic/mbc-mosto/node_modules/mbc-common/node_modules/mongoskin/node_modules/mongodb/lib/mongodb/cursor.js:155:9
      at /home/travis/build/inaes-tic/mbc-mosto/node_modules/mbc-common/node_modules/mongoskin/node_modules/mongodb/lib/mongodb/cursor.js:195:11
      at /home/travis/build/inaes-tic/mbc-mosto/node_modules/mbc-common/node_modules/mongoskin/node_modules/mongodb/lib/mongodb/cursor.js:545:35
      at Cursor.close (/home/travis/build/inaes-tic/mbc-mosto/node_modules/mbc-common/node_modules/mongoskin/node_modules/mongodb/lib/mongodb/cursor.js:776:5)
      at Cursor.nextObject (/home/travis/build/inaes-tic/mbc-mosto/node_modules/mbc-common/node_modules/mongoskin/node_modules/mongodb/lib/mongodb/cursor.js:545:17)
      at commandHandler (/home/travis/build/inaes-tic/mbc-mosto/node_modules/mbc-common/node_modules/mongoskin/node_modules/mongodb/lib/mongodb/cursor.js:522:14)
      at g (events.js:175:14)
      at EventEmitter.emit (events.js:106:17)
      at Db._callHandler (/home/travis/build/inaes-tic/mbc-mosto/node_modules/mbc-common/node_modules/mongoskin/node_modules/mongodb/lib/mongodb/db.js:1472:25)
      at /home/travis/build/inaes-tic/mbc-mosto/node_modules/mbc-common/node_modules/mongoskin/node_modules/mongodb/lib/mongodb/connection/server.js:430:30
      at MongoReply.parseBody (/home/travis/build/inaes-tic/mbc-mosto/node_modules/mbc-common/node_modules/mongoskin/node_modules/mongodb/lib/mongodb/responses/mongo_reply.js:127:5)
      at null.<anonymous> (/home/travis/build/inaes-tic/mbc-mosto/node_modules/mbc-common/node_modules/mongoskin/node_modules/mongodb/lib/mongodb/connection/server.js:389:22)
      at EventEmitter.emit (events.js:95:17)
      at null.<anonymous> (/home/travis/build/inaes-tic/mbc-mosto/node_modules/mbc-common/node_modules/mongoskin/node_modules/mongodb/lib/mongodb/connection/connection_pool.js:140:13)
      at EventEmitter.emit (events.js:98:17)
      at Socket.<anonymous> (/home/travis/build/inaes-tic/mbc-mosto/node_modules/mbc-common/node_modules/mongoskin/node_modules/mongodb/lib/mongodb/connection/connection.js:321:22)
      at Socket.EventEmitter.emit (events.js:95:17)
      at Socket.<anonymous> (_stream_readable.js:736:14)
      at Socket.EventEmitter.emit (events.js:92:17)
      at emitReadable_ (_stream_readable.js:408:10)
      at emitReadable (_stream_readable.js:404:5)
      at readableAddChunk (_stream_readable.js:165:9)
      at Socket.Readable.push (_stream_readable.js:127:10)
      at TCP.onread (net.js:526:21)
```

Lacrymology commented:
I understand how this failed.. what I don't understand is why it didn't fail for me locally...

Lacrymology commented:
I rebased this onto the new master

jmrunge commented:
I restarted build on Travis.  Still failing.  It also fails in my machine.  Same error.  Can you please take another look at this?

Lacrymology commented:
I am. I've been on this for two days, and I cannot make it work properly.

jmrunge commented:
I remember having this problem when reviewing @josx's PR from Backbone Relational.  I still have this branch local and this is how createPlaylist function looked:

``` javascript
mongo_driver.prototype.createPlaylist = function(sched, callback) {
    var self = this;
    var err = undefined;
    console.log("mongo-driver: [INFO] Create Playlist:", sched);

    var startDate = moment.unix(sched.start).valueOf();
    var endDate = moment.unix(sched.end).valueOf();
    var name = sched.title;
    var playlist_id = (sched._id.toHexString && sched._id.toHexString()) || sched._id;

    var medias = [];
    _.forEach(sched.list.pieces, function(piece, order) {
        var piece_id = (piece._id.toHexString && piece._id.toHexString()) || piece._id;
        var orig_order = order;
        var clip_name = piece.name;
        // TODO: don't know what goes in type
        var type = "default";
        var file = piece.file;
        var length = moment(piece.durationraw, "HH:mm:ss.SSS");
        var fps = piece.video.fps;
        medias.push(new Media(piece_id, orig_order, playlist_id, clip_name, type, file,
                              moment.duration({
                                  hours: length.hours(),
                                  minutes: length.minutes(),
                                  seconds: length.seconds(),
                                  milliseconds: length.milliseconds(),
                              }), parseFloat(fps)));
    });

    var playlist = new Playlist(playlist_id, name, startDate, medias, endDate, "snap");

    callback(err, playlist);

};
```

Lacrymology committed:
create test db data from models instead of hardcoded file

Signed-off-by: Tomas Neme <lacrymology@gmail.com>

Lacrymology committed:
create pubsub message from models, too

Signed-off-by: Tomas Neme <lacrymology@gmail.com>

Lacrymology commented:
For this to work, https://github.com/inaes-tic/mbc-common/pull/20 needs to be merged.

Lacrymology commented:
I can't reproduce the current travis failure....

jmrunge commented:
Im having the exact same error as travis does.  Deleted mbc-common and
backbone node_module's dirs and npm installed them.

2013/10/9 Tomas Neme notifications@github.com

> I can't reproduce the current travis failure....
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/inaes-tic/mbc-mosto/pull/123#issuecomment-26007561
> .

Lacrymology commented:
did you do that since yesterday? because we just merged some changes into
common

jmrunge committed:
Left comented code for @Lacrymology to take a look at and then erase it!

Signed-off-by: Juan Martin Runge <jmrunge@gmail.com>

jmrunge committed:
Restored skipped test

Signed-off-by: Juan Martin Runge <jmrunge@gmail.com>

jmrunge committed:
Comented setWindow code as it does not exists any more

Signed-off-by: Juan Martin Runge <jmrunge@gmail.com>

jmrunge committed:
Added title attr to playlist so log message stops giving error

Signed-off-by: Juan Martin Runge <jmrunge@gmail.com>

jmrunge committed:
Added setTimeout after saving objects to db so test stop failing when trying to recover same object

Signed-off-by: Juan Martin Runge <jmrunge@gmail.com>

jmrunge committed:
Changed playlist start test as attr startDate didnt exist any more and start attr already is a moment

Signed-off-by: Juan Martin Runge <jmrunge@gmail.com>

jmrunge commented:
Of course I tested this after mbc-common merge.  Please take a look at my commits.  Test is passing now locally on my machine.

Lacrymology committed:
remove comments

Signed-off-by: Tomas Neme <lacrymology@gmail.com>

Lacrymology committed:
save with {safe:true}

unless you pass {safe:true} to mongodb-native's insert/update/save, the callback is fired immediatly

http://mongodb.github.io/node-mongodb-native/markdown-docs/insert.html

Signed-off-by: Tomas Neme <lacrymology@gmail.com>
